src/sparseml/optim/helpers.py

```diff
@@ -438,3 +444,86 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
             return float(val)
         except Exception:
             return val
+
+
+def validate_metadata(metadata: dict, yaml_str: str) -> dict:
```

Comment:
let's redo the logic here with the following rule:

1. if no metadata if found in the `yaml_str` then all is good
2. if metadata if found in the `yaml_str` and is equal to the given `metadata` dict, then all is good
3. if the given `metadata` and `yaml_str` metadata conflict we raise a warning and possible exception

in other words, we should never merge, because the metadata for a previous run has no bearing on the current run and if the metadata differs then we want the user to know that they might not be able to reproduce that same exact result (of course, that may be the goal :) )